File: quick-sort/quick-sort.py
nums = [100,88,77,55,33,22,75,58,1]

def quick_sort(alist, low, high):
  if low >= high:
    return
  idx = weak_sort(alist, low, high)

  quick_sort(alist, low, idx - 1)
  quick_sort(alist, idx + 1, high)


def weak_sort(alist, low, high):
  pivot = (high + low) // 2 + 1
  idx = low - 1
  print(idx, alist)
  for i in range(low, high + 1):
    if pivot != i and alist[i] <= alist[pivot]:
      idx += 1
      alist[i], alist[idx] = alist[idx], alist[i]
  idx += 1
  alist[idx], alist[pivot] = alist[pivot], alist[idx]
  return idx

quick_sort(nums, 0, len(nums)-1)
print(nums)

File: quick-sort/quickSort.ts
const nums = [100,88,77,66,55,44,33,85,74,49,1]

quickSort(nums,0,nums.length - 1)
console.log(nums)

function quickSort(nums: number[], low: number, high: number):void{
  
  if (low < high){
    const pivot = weakSort(nums,low,high)
    quickSort(nums,low, pivot - 1)
    quickSort(nums, pivot + 1, high)  
  }
}

function weakSort(nums: number[], low: number, high: number){
  const pivot = Math.floor((high + low) / 2) + 1
  const current = nums[pivot]
  for(let i = low; i <= high; i++){
    if(nums[i] < current){
      const temp = nums[i]
      nums[i] = nums[low]
      nums[low] = temp
      low++
    }
  }
  nums[pivot] = nums[low]
  nums[low] = current
  return low
}